packages/editor/src/lib/editor/managers/HistoryManager.ts

```diff
@@ -1,156 +1,120 @@
+		'HistoryManager.stacks',
+		{
+			undos: stack<TLHistoryEntry<R>>(),
+			redos: stack<TLHistoryEntry<R>>(),
+		},
+		{
+			isEqual: (a, b) => a.undos === b.undos && a.redos === b.redos,
 		}
-		this._commands[name] = handle
-
-		const exec = (...args: ExtractArgs<Constructor>) => {
-			if (!this._batchDepth) {
-				// If we're not batching, run again in a batch
-				this.batch(() => exec(...args))
-				return this.ctx
-			}
-
-			const result = constructor(...args)
-
-			if (!result) {
-				return this.ctx
+	)
+
+	private readonly annotateError: (error: unknown) => void
+
+	constructor(opts: { store: Store<R>; annotateError?: (error: unknown) => void }) {
+		this.store = opts.store
+		this.annotateError = opts.annotateError ?? noop
+		this.dispose = this.store.addHistoryInterceptor((entry, source) => {
+			if (source !== 'user') return
+
+			switch (this.state) {
+				case HistoryRecorderState.Recording:
+					this.pendingDiff.apply(entry.changes)
+					this.stacks.update(({ undos }) => ({ undos, redos: stack() }))
+					break
+				case HistoryRecorderState.RecordingPreserveRedoStack:
+					this.pendingDiff.apply(entry.changes)
+					break
+				case HistoryRecorderState.Paused:
+					break
+				default:
+					exhaustiveSwitchError(this.state)
 			}
+		})
+	}
 
-			const { data, ephemeral, squashing, preservesRedoStack } = result
-
-			this.ignoringUpdates((undos, redos) => {
-				handle.do(data)
-				return { undos, redos }
-			})
-
-			if (!ephemeral) {
-				const prev = this._undos.get().head
-				if (
-					squashing &&
-					prev &&
-					prev.type === 'command' &&
-					prev.name === name &&
-					prev.preservesRedoStack === preservesRedoStack
-				) {
-					// replace the last command with a squashed version
-					this._undos.update((undos) =>
-						undos.tail.push({
-							...prev,
-							data: devFreeze(handle.squash!(prev.data, data)),
-						})
-					)
-				} else {
-					// add to the undo stack
-					this._undos.update((undos) =>
-						undos.push({
-							type: 'command',
-							name,
-							data: devFreeze(data),
-							preservesRedoStack: preservesRedoStack,
-						})
-					)
-				}
-
-				if (!result.preservesRedoStack) {
-					this._redos.set(stack())
-				}
+	private flushPendingDiff() {
+		if (this.pendingDiff.isEmpty()) return
 
-				this.ctx.emit('change-history', { reason: 'push' })
-			}
+		const diff = this.pendingDiff.clear()
+		this.stacks.update(({ undos, redos }) => ({
+			undos: undos.push({ type: 'diff', diff }),
+			redos,
+		}))
+	}
 
-			return this.ctx
-		}
+	onBatchComplete: () => void = () => void null
 
-		return exec
+	getNumUndos() {
+		return this.stacks.get().undos.length + (this.pendingDiff.isEmpty() ? 0 : 1)
 	}
+	getNumRedos() {
+		return this.stacks.get().redos.length
+	}
+
+	/** @internal */
+	_isInBatch = false
+	batch = (fn: () => void, opts?: TLHistoryBatchOptions) => {
+		const previousState = this.state
+		this.state = opts?.history ? modeToState[opts.history] : this.state
```

Comment:
As we discussed in our sync the other day, I think `ignore` should probably overrule any nested usages of history.batch with other values.
```suggestion
		if (this.state !== HistoryRecorderState.Paused) {
			this.state = opts?.history ? modeToState[opts.history] : this.state
		}
```